test/models/bot/tagger_test.rb

```diff
@@ -0,0 +1,117 @@
+require_relative '../../test_helper'
+
+class Bot::TaggerTest < ActiveSupport::TestCase
+  def setup
+    super
+    @TaggerBot = create_tagger_bot(name: "tagger", login: "tagger")
+    @AlegreBot = create_alegre_bot(name: "alegre", login: "alegre")
+    @TaggerBot.approve!
+    @AlegreBot.approve!
+    p = create_project
+    p.team.save!
+    @AlegreBot.install_to!(p.team)
+    @TaggerBot.install_to!(p.team)
+    @team = p.team
+    m = create_claim_media quote: 'I like apples'
+    @pm = create_project_media project: p, media: m
+    create_flag_annotation_type
+    create_extracted_text_annotation_type
+    Sidekiq::Testing.inline!
+    @auto_tag_prefix = "!"
+  end
+
+  def teardown
+    super
+  end
+
+  def create_tagger_bot(_options = {})
+    Bot::Tagger.new(_options)
+  end
+
+  test "should get tag text" do
+    # Method signature
+    # get_tag_text(tag_id,auto_tag_prefix,ignore_autotags)
+    pm1 = create_project_media quote: "Rain", team: @team
+    tag1 = Tag.create!(annotated:pm1, annotator: BotUser.get_user('tagger'), tag: @auto_tag_prefix+"test")
+
+    pm2 = create_project_media quote: "Rain", team: @team
+    tag2 = Tag.create!(annotated:pm2, annotator: BotUser.get_user('tagger'), tag: "test")
+
+    # Get tag text keeping autotags
+    # Should return tag with prefix stripped
+    assert_equal "test", Bot::Tagger.get_tag_text(tag1[:data][:tag], @auto_tag_prefix, false)
+
+    # Get tag text ignoring autotags
+    assert_nil Bot::Tagger.get_tag_text(tag1[:data][:tag], @auto_tag_prefix, true)
+    assert_equal "test", Bot::Tagger.get_tag_text(tag2[:data][:tag], @auto_tag_prefix, true)
+
+    # Get tag text for invalid id
+    # Should return nil
+    assert_nil Bot::Tagger.get_tag_text(-999, @auto_tag_prefix, false)
+  end
+
+  test "should tag item" do
+
+  pm1 = create_project_media quote: "rain", team: @team
+  Tag.create!(annotated:pm1, annotator: BotUser.get_user('tagger'), tag: "nature")
+  pm2 = create_project_media quote: "snow", team: @team
+  Tag.create!(annotated:pm2, annotator: BotUser.get_user('tagger'), tag: @auto_tag_prefix+"nature")
+  pm3 = create_project_media quote: "soccer", team: @team
+  Tag.create!(annotated:pm3, annotator: BotUser.get_user('tagger'), tag: "sport")
+
+  #Stub alegre to return pm1, pm2, and pm3 for any request
+  Bot::Alegre.stubs(:get_items_with_similar_text).returns({pm1.id=>pm1,pm2.id=>pm2,pm3.id=>pm3})
```

Comment:
What's best practice here @caiosba ? Should I stub the AlegreBot method I use or the go all the way down to the :request_api method where AlegreBot would actually call the Alegre service

Reply:
@computermacgyver best practice is to stub the highest level you're not testing... so you don't need to go all way down to the low-level `request_api` method.